lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp

```diff
@@ -1218,164 +1211,207 @@ namespace Js
             {
-                aLeft = const_cast<char16*>(str1->GetSz());
-                size1 = str1->GetLength();
+                locale = defaultLocale;
             }
-            if (aRight == nullptr)
+            else
             {
-                aRight = const_cast<char16*>(str2->GetSz());
-                size2 = str2->GetLength();
+    #if defined(INTL_WINGLOB)
+                // win32 GetUserDefaultLocaleName returns its error through GetLastError
+                JavascriptError::MapAndThrowError(scriptContext, HRESULT_FROM_WIN32(GetLastError()));
+    #elif defined(INTL_ICU)
+                // TODO(jahorto): what error should we throw here for INTL_ICU?
+                JavascriptError::MapAndThrowError(scriptContext, E_FAIL);
+    #endif
             }
+        }
 
-#ifdef _WIN32
-            // xplat-todo: Need to replace this with platform-agnostic API
-            compareResult = CompareStringEx(givenLocale != nullptr ? givenLocale : defaultLocale, compareFlags, aLeft, size1, aRight, size2, NULL, NULL, 0);
-#else
-            // FIXME (doilij): when CompareStringEx is implemented in PlatformAgnostic, call that function here
-            compareResult = 2; // 2 means strings are equal (reasonable default)
-#endif
+        Assert(locale != nullptr);
+        Assert(sensitivity >= CollatorSensitivity::DEFAULT && sensitivity < CollatorSensitivity::MAX);
+        Assert(caseFirst >= CollatorCaseFirst::DEFAULT && caseFirst < CollatorCaseFirst::MAX);
 
-            // Get the last error code so that it won't be affected by END_TEMP_ALLOCATOR.
-            if (compareResult == 0)
-            {
-                lastError = GetLastError();
-            }
+        BEGIN_TEMP_ALLOCATOR(tempAllocator, scriptContext, _u("EntryIntl_CompareString"));
+
+        const char16 *normalized1 = nullptr;
+        charcount_t normalized1Length = 0;
+        if (UnicodeText::IsNormalizedString(UnicodeText::NormalizationForm::C, str1->GetSz(), str1->GetLength()))
+        {
+            normalized1 = str1->GetSz();
+            normalized1Length = str1->GetLength();
         }
+        else
+        {
+            normalized1 = str1->GetNormalizedString(UnicodeText::NormalizationForm::C, tempAllocator, normalized1Length);
+        }
+
+        const char16 *normalized2 = nullptr;
+        charcount_t normalized2Length = 0;
+        if (UnicodeText::IsNormalizedString(UnicodeText::NormalizationForm::C, str2->GetSz(), str2->GetLength()))
+        {
+            normalized2 = str2->GetSz();
+            normalized2Length = str2->GetLength();
+        }
+        else
+        {
+            normalized2 = str2->GetNormalizedString(UnicodeText::NormalizationForm::C, tempAllocator, normalized2Length);
+        }
+
+        // CompareStringEx on Windows returns 0 for error, 1 if less, 2 if equal, 3 if greater
+        // Default to the strings being equal
+        int compareResult = 2;
```

Comment:
Default is equal?

Reply:
Yeah because if you have a < b and b < a then you can never sort something (no convergence, infinite loop). Everything equal means no sorting but convergence.

---
In reply to: [148364888](https://github.com/Microsoft/ChakraCore/pull/4083#discussion_r148364888) [](ancestors = 148364888)